pkg/sentry/kernel/fd_table.go

```diff
@@ -267,31 +267,56 @@ func (f *FDTable) NewFDs(ctx context.Context, fd int32, files []*fs.File, flags
 		if lim.Cur != limits.Infinity {
 			end = int32(lim.Cur)
 		}
-		if fd >= end {
+		if minFD >= end {
 			return nil, unix.EMFILE
 		}
 	}
 
 	f.mu.Lock()
 
-	// From f.next to find available fd.
-	if fd < f.next {
-		fd = f.next
+	// max is the largest bit in fdBitmap.
+	max := int32(0)
+	// flipFdBitmap is flip of fdBitmap which is used to find free fd in fdBitmap.
+	flipFdBitmap := f.fdBitmap.Clone()
+
+	if !f.fdBitmap.IsEmpty() {
+		max = int32(f.fdBitmap.Maximum())
+		flipFdBitmap.FlipRange(uint32(0), uint32(max))
```

Comment:
After moving the 2 above lines directly below the declaration of flipFdBitmap, the 0 bit would be flipped when fdBitmap is empty. This would lead to an error that FD 0 would be used even if the `minfd` is larger than 0. To fix this, here is a draft code.

```
@@ -278,16 +278,19 @@ func (f *FDTable) NewFDs(ctx context.Context, minFD int32, files []*fs.File, fla
        max := int32(0)
        // flipFdBitmap is flip of fdBitmap which is used to find free fd in fdBitmap.
        flipFdBitmap := f.fdBitmap.Clone()
+       max = int32(f.fdBitmap.Maximum())
+       flipFdBitmap.FlipRange(uint32(0), uint32(max))

        if !f.fdBitmap.IsEmpty() {
-               max = int32(f.fdBitmap.Maximum())
-               flipFdBitmap.FlipRange(uint32(0), uint32(max))
                max++
                // Clear 0 to (minFD-1) bit in flipFdBitmap, thus it will not use fd that is less than minFD.
                if minFD > 0 {
                       flipFdBitmap.ClearRange(uint32(0), uint32(minFD - 1))
                }
+       } else {
+               flipFdBitmap.ClearRange(uint32(0), uint32(0))
        }
+
        // Adjust max in case it is less than minFD.
        if max < minFD {
                max = minFD
@@ -296,7 +299,7 @@ func (f *FDTable) NewFDs(ctx context.Context, minFD int32, files []*fs.File, fla
        for len(fds) < len(files) {
                // Try to use free bit in fdBitmap.
                // If all bits in fdBitmap are used, expand fd to the max.
-               if !flipFdBitmap.IsEmpty() {
+               if !f.fdBitmap.IsEmpty() && !flipFdBitmap.IsEmpty() {
                        fd := flipFdBitmap.Minimum()
                        f.fdBitmap.Add(fd)
                        flipFdBitmap.Remove(fd)

```

Reply:
Gotcha. Let's just leave it like it is then.